stellargraph/mapper/graphwave_generator.py

```diff
@@ -0,0 +1,176 @@
+    DataSet that contains the GraphWave embeddings.
+    """
+
+    def __init__(self, G, scales="auto", num_scales=3, num_eigenvecs=-1):
+        """
+        Args:
+            G (StellarGraph): the StellarGraph object.
+            scales (str or list of floats): the wavelet scales to use. "auto" will cause the scale values to be
+                automatically calculated.
+            num_scales (int): the number of scales when scales = "auto".
+            num_eigenvecs (int): the number of eigenvectors to use. When set to `-1` the maximum number of eigenvectors
+                is calculated.
+        """
+
+        if not isinstance(G, StellarGraph):
+            raise TypeError("G must be a StellarGraph object.")
+
+        node_types = list(G.node_types)
+        if len(node_types) > 1:
+            raise TypeError(
+                "{}: node generator requires graph with single node type; "
+                "a graph with multiple node types is passed. Stopping.".format(
+                    type(self).__name__
+                )
+            )
+
+        # Create sparse adjacency matrix:
+        # Use the node orderings the same as in the graph features
+        self.node_list = G.nodes_of_type(node_types[0])
+        self.Aadj = G.to_adjacency_matrix(self.node_list)
+
+        # Function to map node IDs to indices for quicker node index lookups
+        # TODO: Move this to the graph class
+        node_index_dict = dict(zip(self.node_list, range(len(self.node_list))))
+        self._node_lookup = np.vectorize(node_index_dict.get, otypes=[np.int64])
+
+        adj = G.to_adjacency_matrix().tocoo()
+        degree_mat = diags(np.array(adj.sum(1)).flatten())
+        laplacian = degree_mat - adj
+
+        if num_eigenvecs == -1:
+            num_eigenvecs = laplacian.shape[0] - 2
+
+        self.eigen_vals, self.eigen_vecs = eigs(laplacian, k=num_eigenvecs)
+        self.eigen_vals = np.real(self.eigen_vals).astype(np.float32)
+        self.eigen_vecs = np.real(self.eigen_vecs).astype(np.float32)
+
+        if scales == "auto":
+
+            e2 = self.eigen_vals[self.eigen_vals > 0].min()
+            eN = self.eigen_vals.max()
+
+            min_scale = -np.log(0.95) / np.sqrt(eN * e2)
+            max_scale = -np.log(0.85) / np.sqrt(eN * e2)
+
+            scales = np.linspace(min_scale, max_scale, num_scales)
+
+        self.scales = scales
+
+        # the columns of U exp(-scale * eigenvalues) U^T (U = eigenvectors) are used to calculate the node embeddings
+        # (each column corresponds to a node)
+        # to avoid computing a dense NxN matrix when only several eigenvalues are specified
+        # U exp(-scale * eigenvalues) is computed and stored - which is an N x num_eigenvectors matrix
+        # the columns of exp(-scale * eigenvalues) U^T are then computed on the fly in generator.flow()
+        self.Ues = [
+            self.eigen_vecs.dot(np.diag(np.exp(-s * self.eigen_vals))) for s in scales
```

Comment:
Ah I initially discounted this because the paper says "computing each power of the Laplacian is O(E)" which isn't true (usually after a couple of powers L^n becomes dense) but I just had a look at their references (7 and 32) and it seems that the methods proposed there are actually `O(kE)` because they use only use sparse matrix vector products. 

The papers are fairly complicated (at least for me!) so I'll have a closer look at them tomorrow and implement and compare the methods. Reference 7 (https://arxiv.org/pdf/0912.3848.pdf) actually mentions that method (Arnoldi iteration to compute some eigenvalues) as a potential alternative so it could be worth having both. 

btw section 3.1 is talking about something different - its looking at polynomial approximations of `exp(-lambda s)` to justify why the GraphWave encodes structural properties.

Reply:
I wasn't suggesting that we change here, and I think it's better to land in some form and then optimise. In particular, I think we can:

- file an issue
- add a FIXME
- add a note about being O(n^3) (or whatever it is) to the `@experimental` and an issue to the issue

And then follow up later. As you say, this seems like a potentially complicated topic, so having this PR blocked on a deeper investigation of the options seems unfortunate, and, potentially means a pile more only tangentially related code.

Reply:
filed #853

Reply:
Interestingly, the time complexity as is is `O(m^2N + kE + mN^2)` = `O(mN^2)` where `m` is the number of eigenvalues calculated, `E` is the number of edges, and `N` is the number of nodes.

The computational complexity of the Chebyshev approach  is  (I think) `O(kEN)` where `k` is the order of the Chebyshev polynomial, soo this is actually more expensive but I suspect that like in the Watch Your Step case this will be more a lot more accurate than the eigenvalue approach.